Documentation - CI/CD Tutorial End to End with Skaffold (#4909)
GoogleContainerTools/skaffold: docs/content/en/docs/tutorials/ci_cd.md

```

Skaffold can use your dockerfile to automatically build, tag and push images to the registry when needed.

**Step 2: Choosing the deployment method**

Skaffold supports deployment through kubectl, helm and a lot of other mechanisms. You can have a look at the yaml reference for complete list here: [https://skaffold.dev/docs/references/yaml/](https://skaffold.dev/docs/references/yaml/)

So, if you choose to deploy through kubectl you must have all your yaml files ready and if you would like to deploy through helm charts, you must have all your charts ready.

**Step 3: Choosing the image and tag strategy**

Skaffold automatically tags images based on different strategies as documented here: [https://skaffold.dev/docs/pipeline-stages/taggers/](https://skaffold.dev/docs/pipeline-stages/taggers/) and you might want to choose the mechanism which is right for you. By default, it uses the git sha if unspecified.

Also, if you are using helm chart as your deployment method, you might want to use the helm image strategy if you would like to follow helm specific conventions and skaffold will pass in all the details for you.

**Step 4: Plugins (optional)**

> Another interesting thing is that skaffold also supports plugins like Helm Secrets ([https://github.com/zendesk/helm-secrets/](https://github.com/zendesk/helm-secrets/)). So, if you would like to deploy secrets that have been encrypted using KMS or any other mechanism supported by SOPS ([https://github.com/mozilla/sops](https://github.com/mozilla/sops)) you can actually set **useHelmSecrets** option to true and skaffold will handle everything automatically for you.

Reviewer: ```suggestion
Skaffold also supports plugins like Helm Secrets ([https://github.com/zendesk/helm-secrets/](https://github.com/zendesk/helm-secrets/)). So, if you would like to deploy secrets that have been encrypted using KMS or any other mechanism supported by SOPS ([https://github.com/mozilla/sops](https://github.com/mozilla/sops)) you can actually set **useHelmSecrets** option to true and skaffold will handle everything automatically for you.
```

**Step 5: The skaffold.yaml file**

> Depending on the way you would like to build, deploy, your skaffold yaml file will actually change and it's quite easy to set up one. It is recommended that you set up skaffold before you setup your CI pipeline since the CI pipeline can just call skaffold to do all the builds and deployments and you can do the same locally as well.

Reviewer: ```suggestion
Your skaffold.yaml file will change depending on the way you would like to build and deploy your project. It is recommended that you set up skaffold before you setup your CI pipeline since the CI pipeline can just call skaffold to do all the builds and deployments, and you can do the same locally as well.
```

  
*   **Specifying the artifacts to build from** - In this case, we are asking skaffold to build from a Dockerfile, passing its args and also specifying the destination where the image needs to be pushed. The values can also come from the env (NPM_REGISTRY and NPM_TOKEN in this example)

*   **Specifying the deployment strategy** - In this case, helm is specified with the location of the chart, all the configuration as documented in the yaml reference and also the values to be passed to it. We are also using the helm secrets plugin above.

If you want to deploy via kubectl, your skaffold file can look something like this:

```
apiVersion: skaffold/v2beta8
kind: Config
profiles:
    - name: dev-svc
      build:
          artifacts:
              - image: asia.gcr.io/my-project/my-image
                docker:
                    buildArgs:
                        NPM_REGISTRY: '{{.NPM_REGISTRY}}'
                        NPM_TOKEN: '{{.NPM_TOKEN}}'
      deploy:
          statusCheckDeadlineSeconds: 600
          kubectl:
              manifests:
                  - k8/dev/*.yml
```

> **Please refer the yaml reference for all that you can specify in skaffold.yaml file**

Reviewer: ```suggestion
**Please refer to the yaml reference for all that you can specify in skaffold.yaml file**
```

> ## **Step 6: Development & Testing locally**

Reviewer: The labels for steps 1-5 use emphasis, but steps 6-9 use emphasis and are also headers. This styling should be consistent throughout.

Before we move on to setting up the CI pipeline, we should test it out locally. Use **skaffold run** if you are deploying it to the cluster, **skaffold build **to just build and push the artifacts, **skaffold dev **if you are developing (this will enable auto reload on changes) or debug using **skaffold debug**

You can find docs about all these workflows here: [https://skaffold.dev/docs/workflows/](https://skaffold.dev/docs/workflows/)

Also, note that you can also set up file synchronization to enable faster development workflows. You can read more about that here: [https://skaffold.dev/docs/pipeline-stages/filesync/](https://skaffold.dev/docs/pipeline-stages/filesync/)

## Step 7: Getting the CI pipeline setup

Now that we have skaffold ready and working locally, the CI pipeline requires little to no effort since it will just call skaffold to do all the builds and deployments.

> Here is a sample pipeline using Gitlab CI

Author: Hmm.. I understand your point. Since I work with Gitlab for my workflow, I went with this tutorial keeping Gitlab in mind and I hope other CI solutions are similar as well (I guess I may need some external opinion on this). I can add some explanation to this to add more clarity. Regarding REG_DEV_URL, I have mentioned ^^ I had actually replaced the private registry url with gcr since this is a public tutorial. and left REG_DEV_URL as is without removing it. I will replace all instances to suit GCR if that would be fine.

```
stages:
    - my-stage

my-job:
    image:
        name: gcr.io/k8s-skaffold/skaffold:v1.15.0
    tags:
        - development
    stage: my-stage
    retry: 2
    script:

>         - docker login $REG_DEV_URL -u $REG_DEV_UNAME -p $REG_DEV_PASSWORD

Author: @briandealwis Actually, I used to have a private registry (Harbor) and this is how I used to authenticate to it. Since this is going to be a public tutorial, I replaced the image URL with gcr but then left this instruction out as such. This is for passing in user name, password and registry url in case you are either using docker hub or a private registry.

Should I change this to make it suitable to GCR? The process can be different for different registry provider, I thought this would be consistent. Let me know.

    only:
        refs:
            - master
```

Before we look at the pipeline, you might want to look at Gitlab CI/CD environment variables here: [https://docs.gitlab.com/ee/ci/variables/](https://docs.gitlab.com/ee/ci/variables/)

This is how we will configure the various environment variables to be passed in to the pipeline. Make sure that you mask/protect them using Gitlab so that they are not visible in the logs.

Here, we are:

*   Logging into the docker registry specifying the Registry URL, User ID and password (You can choose to authenticate using service tokens as well or other auth providers as well)
*   Logging in to our gcp account using the service account key
*   Specifying the project, zone, cluster to deploy our application in.
*   We are using DIND (Docker In Docker) to facilitate our builds (more about that below)
*   Then we just call the skaffold run in the pipeline passing flags we want to pass at runtime (in our case our cluster context, env vars, namespace and some labels)

> *   And we trigger this only on changes to master branch

Author: @briandealwis Thanks a lot for your suggestions.  Sure. Can change this. Yes, I use a private registry (Verdaccio) and that is what it authenticates with - it can be avoided, but I just put it here so that people would understand how to pass in args from gitlab to skaffold which then gets passed to the build args in Docker.

So, ultimately our Gitlab CI pipeline does not know anything about the helm charts or build strategy or any deployment methods and relies completely on skaffold to do the job for us.

> This works equally well irrespective of whether you choose the Docker Executor ([https://docs.gitlab.com/runner/executors/docker.html](https://docs.gitlab.com/runner/executors/docker.html)) or the Kubernetes Executor ([https://docs.gitlab.com/runner/executors/kubernetes.html](https://docs.gitlab.com/runner/executors/kubernetes.html)) and would require little to no changes.

Reviewer 2: This assumes some knowledge about Gitlab.  I *think* you're using the Kubernetes executor here (since you have a deployment yaml below, which doesn't look related to your application).  Why did you choose this if you could use the Docker Executor?

Author: Docker executor does not run in the Kubernetes cluster, requires access to the Docker engine and also does not allow for things like scaling up/down runners as pods in the cluster all of which is supported in Kubernetes executor. While all the instructions mentioned in this tutorial will work in both runners, Kubernetes executor has its own set of advantages. Should I elaborate about it here?

Reviewer 2: I think the example should use the simplest approach (which I'm guessing is the Docker executor).  You could include a pointer to the Kubernetes executor and briefly summarize some of its advantages.

**NOTE:** If you are running Skaffold using Docker, make sure that you are running the runner with appropriate permissions or pod security policies since running DIND requires access to the Docker socket and being in privileged mode which might actually tend to be insecure. An alternative can be to use kaniko or custom builders like Buildah for building the image.

> ## **Step 9: Using DIND Service (Optional)**

Reviewer: Numbering for steps 8 and 9 should be flipped
